crates/ruma-identifiers/src/matrix_uri.rs

```diff
@@ -234,16 +304,171 @@ impl TryFrom<&str> for MatrixToUri {
     }
 }
 
+/// The intent of a Matrix URI.
+#[derive(Clone, Debug, PartialEq, Eq)]
+#[non_exhaustive]
+pub enum UriAction {
+    /// Join the room referenced by the URI.
+    ///
+    /// The client should prompt for confirmation prior to joining the room, if
+    /// the user isn’t already part of the room.
+    Join,
+
+    /// Start a direct chat with the user referenced by the URI.
+    ///
+    /// Clients supporting a form of Canonical DMs should reuse existing DMs
+    /// instead of creating new ones if available. The client should prompt for
+    /// confirmation prior to creating the DM, if the user isn’t being
+    /// redirected to an existing canonical DM.
+    Chat,
+
+    #[doc(hidden)]
+    _Custom(PrivOwnedStr),
+}
+
+impl UriAction {
+    /// Creates a string slice from this `UriAction`.
+    pub fn as_str(&self) -> &str {
+        self.as_ref()
+    }
+}
+
+impl AsRef<str> for UriAction {
+    fn as_ref(&self) -> &str {
+        match self {
+            UriAction::Join => "join",
+            UriAction::Chat => "chat",
+            UriAction::_Custom(s) => s.0.as_ref(),
+        }
+    }
+}
+
+impl fmt::Display for UriAction {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{}", self.as_ref())?;
+        Ok(())
+    }
+}
+
+impl<T> From<T> for UriAction
```

Comment:
I think this should be replaced by impls on specific types from `std`, like we do for other identifier types. Otherwise, users can't implement `From<MyType> for UriAction` and we might also not be able to add some impl we would like to.

Reply:
The implementation can be done in terms of a generic function with these bounds though.